src/services/utilities.ts

```diff
@@ -2956,5 +3076,211 @@ namespace ts {
+                    !ambientModuleDeclarationsAreEqual(oldSourceFile, newSourceFile)
+                ) {
+                    this.clear();
+                    return true;
+                }
+                usableByFileName = newSourceFile.path;
+                return false;
+            },
+        };
+        if (Debug.isDebugging) {
+            Object.defineProperty(wrapped, "__cache", { get: () => cache });
+        }
+        return wrapped;
+
+        function fileIsGlobalOnly(file: SourceFile) {
+            return !file.commonJsModuleIndicator && !file.externalModuleIndicator && !file.moduleAugmentations && !file.ambientModuleNames;
+        }
+
+        function ambientModuleDeclarationsAreEqual(oldSourceFile: SourceFile, newSourceFile: SourceFile) {
+            if (!arrayIsEqualTo(oldSourceFile.ambientModuleNames, newSourceFile.ambientModuleNames)) {
+                return false;
+            }
+            let oldFileStatementIndex = -1;
+            let newFileStatementIndex = -1;
+            for (const ambientModuleName of newSourceFile.ambientModuleNames) {
+                const isMatchingModuleDeclaration = (node: Statement) => isNonGlobalAmbientModule(node) && node.name.text === ambientModuleName;
+                oldFileStatementIndex = findIndex(oldSourceFile.statements, isMatchingModuleDeclaration, oldFileStatementIndex + 1);
+                newFileStatementIndex = findIndex(newSourceFile.statements, isMatchingModuleDeclaration, newFileStatementIndex + 1);
+                if (oldSourceFile.statements[oldFileStatementIndex] !== newSourceFile.statements[newFileStatementIndex]) {
+                    return false;
+                }
+            }
+            return true;
+        }
+    }
+
+    export function createModuleSpecifierCache(): ModuleSpecifierCache {
+        let cache: ESMap<Path, boolean | readonly ModulePath[]> | undefined;
+        let importingFileName: Path | undefined;
+        const wrapped: ModuleSpecifierCache = {
+            get(fromFileName, toFileName) {
+                if (!cache || fromFileName !== importingFileName) return undefined;
+                return cache.get(toFileName);
+            },
+            set(fromFileName, toFileName, moduleSpecifiers) {
+                if (cache && fromFileName !== importingFileName) {
+                    cache.clear();
+                }
+                importingFileName = fromFileName;
+                (cache ||= new Map()).set(toFileName, moduleSpecifiers);
+            },
+            clear() {
+                cache = undefined;
+                importingFileName = undefined;
+            },
+            count() {
+                return cache ? cache.size : 0;
+            }
+        };
+        if (Debug.isDebugging) {
+            Object.defineProperty(wrapped, "__cache", { get: () => cache });
+        }
+        return wrapped;
+    }
+
+    export function isImportableFile(

```

Comment:
This function is somewhat changed in its move from importFixes.ts. It now takes advantage of the new module specifier cache, and combines the previously separate checks of “can I come up with a module specifier” and “do the package.jsons I can see let me use that module specifier.”